# SEUSS-Analysis
Code to analyze effects of gene overexpression on the transcriptome of human stem cells. The study by Parekh, Wu et al can be found here: [insert paper] . All raw and processed data files are at the following GEO accession: [insert GEO].

## Clustering and regression analysis
Both clustering and regression analysis can take quite a while (8+ hours for large datasets), and so are best run from the command line. You will need the perturbLM package to run this analysis, which can be found at github.com/yanwu2014/perturbLM. 

### Command line parameters
* `[counts_matrix]` is the gene expression counts matrix generated by the 10X Chromium pipeline
* `[genotype_dictionary].csv` is the genotype to cells dictionary generated by https://github.com/yanwu2014/genotyping-matrices
* `[output_file].RData` is the output RData file that will be generated by the analysis
* `[batch_regression]` is whether or to run batch correction, either "T" for true, or "F" for false.

Other important parameters can be found in the scripts themselves. The parameters are set to the values used in the publication

### Usage:
Clustering analysis:
```
Rscript hESC_run_clustering.R [counts_matrix].tsv [genotype_dictionary].csv [output_file].RData [batch_regression]
```

Regression analysis:
```
Rscript hESC_run_regression.R [counts_matrix].tsv [genotype_dictionary].csv [output_file].RData [batch_regression]
```

##  Offline analysis
The higher level analyses (gene modules, etc.) and the dropout analyses are relatively quick and are best if used with RStudio: https://www.rstudio.com/.

### Genotype trimming (optional)
After running clustering and regression analyses on the desired sample, you can optionally clean up the genotypes with `hESC_genotype_trimming.R`. Change the parameter `file.handle` to the name of the sample to be analyzed.

### Summarizing differential expression and clustering.
Run the script `hESC_clustering_regression_analysis.R` with the `file.handle` parameter set to the name of the sample to be analyzed. This section will generate a summary table for each sample, as well as differential expression heatmaps, cluster enrichment heatmaps, and tSNE plots.

### Dropout Analyses
All of the dropout analyses in `hESC_dropout_analysis.R` can be run after generating the summary tables for each sample. This script will generate heatmaps showing fitness change across all three media conditions as calculated from genomic DNA reads, and from scRNA-seq cell counts. It will also generate scatterplots showing the correlation between genomic DNA fitness and scRNA-seq cell counts fitness, and scatterplots showing correlation between genomic DNA fitness and the number of significantly differentially expressed genes for each transcription factor.

### Co-perturbation network and gene module analysis
The code in `hESC_coperturbation_network.R` generates and clusters the gene-gene co-perturbation network. It can be run after generating the summary tables for each of the media condition samples.

### Orthologous mouse TF enrichment
Run the script `hESC_mouse_ortholog_comparison.R`. Compares the differentially expressed genes from our screen to the screen performed by Nishiyama et al (https://www.ncbi.nlm.nih.gov/pubmed/19796622) by mapping mouse genes to human orthologs and running a GSEA enrichment.

### Epithelial-mesenchymal transition (EMT) analysis
Run the script `hESC_EMT_analysis.R`. Epithelial mesenchymal transition (EMT) analysis. Runs PCA on known EMT genes and creates a PCA plot showing the separation of KLF4 and SNAI2 cells. Also generates a barplot showing the up/downregulation of canonical EMT markers for KLF4 and SNAI2.

### Regression analysis summary across media conditions
Run `hESC_diff_genes_barplot.R` to generate a stacked barplot summarizing differential expression results across the three media conditions.

### Correlate screens to establish reproducibility
Run `hESC_dataset_correlation.R` to correlate different screens and demonstrate reproducibility. There are 5 different screens for the hPSC media condition, 1 screen for endothelial media, and 2 screens for multilineage media.

### MYC mutant & KLF family analysis
Run `hESC_KLF_MYC_screen_analysis.R` to analyze the KLF family and MYC mutant screens.

### Highmoi screen analysis
Run `hESC_neuron_highmoi_analysis.R` to analyze the neuronal TF highmoi screen
