# SEUSS-Analysis
Code to analyze effects of gene overexpression on the transcriptome of human stem cells. The study by Parekh, Wu et al can be found here: [insert paper] . All raw and processed data files are at the following GEO accession: [insert GEO].

## Clustering and regression analysis
Both clustering and regression analysis can take quite a while (8+ hours for large datasets), and so are best run from the command line.

### Command line parameters
* `[counts_matrix]` is the gene expression counts matrix generated by the 10X Chromium pipeline
* `[genotype_dictionary].csv` is the genotype to cells dictionary generated by https://github.com/yanwu2014/genotyping-matrices
* `[output_file].RData` is the output RData file that will be generated by the analysis
* `[batch_regression]` is whether or to run batch correction, either "T" for true, or "F" for false.

Other important parameters can be found in the scripts themselves. The parameters are set to the values used in the publication

### Usage:
Clustering analysis:
```
Rscript hESC_cluster_analysis.R [counts_matrix].tsv [genotype_dictionary].csv [output_file].RData [batch_regression]
```

Regression analysis:
```
Rscript hESC_cluster_analysis.R [counts_matrix].tsv [genotype_dictionary].csv [output_file].RData [batch_regression]
```

##  Offline analysis
The higher level analyses (gene modules, etc.) and the dropout analyses are relatively quick and are best if used with RStudio: https://www.rstudio.com/. 

### Summarizing differential expression and clustering.
